c_src/notif_set_tlh_htable.c

```diff
@@ -0,0 +1,245 @@
+    if((itr->itr2 = htable_next((struct htable *) itr->itr1->value, itr->itr2))) {
+        assert(itr->itr1->key->n == sizeof(uint64_t));
+        msg->wd = *((uint64_t*) itr->itr1->key->bytes); //Really don't like this
+        memcpy(msg->name, itr->itr2->key->bytes, itr->itr2->key->n);
+        return true;
+    }
+    // If second level is exhausted, advance the iterator in the first level
+    if((itr->itr1 = htable_next(&itr->tab->lev1, itr->itr1))) {
+        // Call recursively to iterate in the second level
+        return notif_set_itr_next(itr, msg);
+    }
+
+    return false;
+}
+
+bool notif_set_itr_remove(const struct Notif_set_itr *itr)
+{
+    // First, attempt to remove from the second-level table
+    if(itr->itr2) {
+        if(htable_key_free((struct Key*) itr->itr2->key)) {
+            itr->itr2->key = NULL;
+            ((struct htable*)(itr->itr1->value))->len--;
+            // TODO: We could destroy the level2 htable and remove
+            //       the corresponding entry in the level1 table if 
+            //       the level2 table becomes empty. But, let's do it 
+            //       in a lazy fashion and leave it to the destroy function.
+            itr->tab->len--;
+            return true;
+        }
+    }
+
+    // If itr2 is NULL, then itr1 must be NULL too, otherwise
+    // we have some kind of a dangling iterator. Of course, I'm
+    // assuming that the iterator is only progressed by _next function.
+    assert(itr->itr1 == NULL);
+    return false;
+}
+
+void notif_set_fprint(FILE *fp, const struct Notif_set *tab)
+{
+    struct htable_bucket *itr1, *itr2;
+    int entry_idx = 0;
+
+    for(itr1 = htable_next(&tab->lev1, NULL); 
+        itr1; 
+        itr1 = htable_next(&tab->lev1, itr1)) {
+        for(itr2 = htable_next((struct htable*) itr1->value, NULL); 
+            itr2; 
+            itr2 = htable_next((struct htable*) itr1->value, itr2)) {
+            fprintf(fp, "Entry %d: WD = %d, name = %s\n", entry_idx, 
+                    *((int*) itr1->key->bytes), 
+                    itr2->key->bytes); //no cast; bytes is char[]
+            entry_idx++;
+        }
+    }
+
+    if(entry_idx == 0) {
+        fprintf(fp, "Table empty!\n");
+    }
+}
+
+void notif_set_print(const struct Notif_set *tab)
+{
+    return notif_set_fprint(stdout, tab);
+}
+
+size_t notif_set_get_length(const struct Notif_set *tab)
+{
+    return tab->len;
+}
+
+bool notif_set_del(struct Notif_set *tab, const struct Message *msg, int *err_flag)
+{
+    *err_flag = 0;
+    struct Key *key = htable_key_int(msg->wd);
+    CHECK_SUCCESS(key);
+    struct htable_ret get_ret = htable_get(&tab->lev1, key);
+    htable_key_free(key);
+    if(!get_ret.ok) {
+        return false;
+    }
+    key = htable_key_str(msg->name);
+    CHECK_SUCCESS(key);
+    if(htable_del((struct htable*) get_ret.value, key).ok) {
+        tab->len--;
+        // TODO: We could destroy the level2 htable and remove
```

Comment:
Again, it might be good to remove the TODO here.

Reply:
Thank you very much for your comments, Richard. I updated things accordingly.